tests/functional/directives.ts

```diff
@@ -0,0 +1,655 @@
+        const mutation = `mutation {
+          mutationWithUpper
+        }`;
+
+        const { data } = await graphql(schema, mutation);
+
+        expect(data).toHaveProperty("mutationWithUpper");
+        expect(data.mutationWithUpper).toBe("MUTATIONWITHUPPER");
+      });
+
+      it("calls directive 'upper' using SDL", async () => {
+        const mutation = `mutation {
+          mutationWithUpperSDL
+        }`;
+
+        const { data } = await graphql(schema, mutation);
+
+        expect(data).toHaveProperty("mutationWithUpperSDL");
+        expect(data.mutationWithUpperSDL).toBe("MUTATIONWITHUPPER");
+      });
+
+      it("calls directive 'append'", async () => {
+        const mutation = `mutation {
+          mutationWithAppend(append: ", world!")
+        }`;
+
+        const { data } = await graphql(schema, mutation);
+
+        expect(data).toHaveProperty("mutationWithAppend");
+        expect(data.mutationWithAppend).toBe("hello, world!");
+      });
+
+      it("calls directive 'append' using SDL", async () => {
+        const mutation = `mutation {
+          mutationWithAppendSDL(append: ", world!")
+      }`;
+
+        const { data } = await graphql(schema, mutation);
+
+        expect(data).toHaveProperty("mutationWithAppendSDL");
+        expect(data.mutationWithAppendSDL).toBe("hello, world!");
+      });
+
+      it("calls directive 'upper' and 'append'", async () => {
+        const mutation = `mutation {
+          mutationWithUpperAndAppend(append: ", world!")
+        }`;
+
+        const { data } = await graphql(schema, mutation);
+
+        expect(data).toHaveProperty("mutationWithUpperAndAppend");
+        expect(data.mutationWithUpperAndAppend).toBe("HELLO, WORLD!");
+      });
+
+      it("calls directive 'upper' and 'append' using SDL", async () => {
+        const mutation = `mutation {
+          mutationWithUpperAndAppendSDL(append: ", world!")
+      }`;
+
+        const { data } = await graphql(schema, mutation);
+
+        expect(data).toHaveProperty("mutationWithUpperAndAppendSDL");
+        expect(data.mutationWithUpperAndAppendSDL).toBe("HELLO, WORLD!");
+      });
+    });
+
+    describe("InputType", () => {
+      it("adds field directive to input types", async () => {
+        const inputType = schema.getType("DirectiveOnClassInput") as GraphQLInputObjectType;
+
+        expect(inputType).toHaveProperty("astNode");
+        assertValidFieldDirective(inputType.astNode, "upper");
+      });
+
+      it("adds field directives to input type fields", async () => {
+        const fields = (schema.getType(
+          "DirectiveOnFieldInput",
+        ) as GraphQLInputObjectType).getFields();
+
+        expect(fields).toHaveProperty("append");
+        expect(fields.append).toHaveProperty("astNode");
+        assertValidFieldDirective(fields.append.astNode, "upper");
+      });
+    });
+
+    describe("ObjectType", () => {
+      it("calls object type directives", async () => {
```

Comment:
do we have tests for queries?

Reply:
Yup, all above.  Queries, Mutations, ObjectTypes and InputTypes atm.